### PR TITLE
Add Github pages deploy

### PR DESCRIPTION
**Pull request by SaraVieira**
- [x] Add UI
- [x] Make work
- [x] Make work when user has already a GH Repo
- [x] Make Sure templates deploy to it proprely
- [x] Polish

### CODE REVIEW

**Comment by lbogdan** [comment]
Build for latest commit https://github.com/codesandbox/codesandbox-client/pull/5561/commits/d85bdf75bc7dd306b3c2c315fd3b78ab0c10d7e1 is at https://pr5561.build.csb.dev/s/new.

File: packages/app/src/app/overmind/utils/items.ts [hunk 1]
```diff
@@ -59,6 +59,7 @@ export const DEPLOYMENT: INavigationItem = {
   id: 'deploy',
   name: 'Deployment',
   showAsDisabledIfHidden: true,
+  defaultOpen: true,
```
Comment (SaraVieira) [inline packages/app/src/app/overmind/utils/items.ts:62]: CHANGE BEFORE MERGE

File: packages/app/src/app/overmind/onInitialize.ts [hunk 2]
```diff
@@ -1,136 +0,0 @@
-import { Context } from '.';
```
Comment (SaraVieira) [inline packages/app/src/app/overmind/onInitialize.ts:1]: @christianalfoni This can be deleted right?

**Review by DannyRuchtie — CHANGES_REQUESTED** [review]
@SaraVieira is it possible to add a spinner? 
Or to open the logs after deployment? 

It can take a while for a project to build and deployed. The first time I deployed and opened the link I got a 404. Github was not ready yet and they did not show a message that it is building the project when you visit too early

**Comment by DannyRuchtie** [comment]
When using assets in a react project do I need to set a different root? 
Deployer this sandbox: https://pr5561.build.csb.dev/s/danny-ruchtie-v6-nottb
This is the result:  https://dannyruchtie.github.io/photography

Not sure that is something I need to change

**Comment by SaraVieira** [comment]
Hey!

The modal will now open with the logs so it's more visible how the deploy is doing and the images issue is fixed.

The issue in yours is mostly to do with our they are being imported like `"/work/Work-00002.jpg"` this will always send you to the root of the website so it needs to be imported like: `"./work/Work-00002.jpg"`

File: packages/app/src/app/overmind/effects/deployment/githubPages.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      _options = options;
+    },
+    async provideJwtCached() {
+      if (!_jwtToken) {
+        try {
+          const token = await _options.provideJwtToken();
+          setTimeout(() => {
+            // Token expires after 10 seconds, for safety we actually cache the token
+            // for 5 seconds
+            _jwtToken = undefined;
```
Comment (CompuIves) [inline packages/app/src/app/overmind/effects/deployment/githubPages.ts:32]: Nice that you were mindful of this!!

File: packages/app/src/app/overmind/effects/deployment/githubPages.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      };
+
+      const env = (name: string) => {
+        if (name === 'create-react-app') {
+          return `PUBLIC_URL=https://${username}.github.io/csb-${sandbox.id}/`;
+        }
+        return '';
+      };
+
+      const website = await axios.post(
```
Comment (CompuIves) [inline packages/app/src/app/overmind/effects/deployment/githubPages.ts:98]: Not related to this PR, but important to evaluate anyway. @andrasbacsai is the builder isolated from everything? These are commands that can break out of the system, essentially remote code execution as a service (I know we're good at this :) ), but we need to be extra mindful that people will be able to run anything. And we need to make sure that people can't break out and read files in if private sandboxes are deployed with this service.

File: packages/app/src/app/overmind/effects/deployment/netlify.ts [hunk 5]
```diff
@@ -7,7 +7,7 @@ const NetlifyBaseURL = 'https://builder.csbops.io/netlify/site';
 
 type Options = {
   getUserId(): string | null;
-  provideJwtToken: () => any;
+  provideJwtToken: () => Promise<string | null>;
```
Comment (CompuIves) [inline packages/app/src/app/overmind/effects/deployment/netlify.ts:10]: Another `any` down the drain! ❤️

File: packages/app/src/app/overmind/namespaces/deployment/actions.ts [hunk 6]
```diff
@@ -219,3 +222,53 @@ export const aliasDeployment = async (
...
+    await effects.githubPages.deploy(
+      sandbox,
+      state.deployment.githubSite.ghLogin
+    );
+    state.deployment.deploying = false;
+
+    state.deployment.building = true;
+    await effects.githubPages.getLogs(sandbox.id);
+    state.deployment.githubSite.ghPages = true;
+    effects.notificationToast.success('Sandbox Deploying');
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/deployment/actions.ts:249]: So If I understand correctly it's deploying at this point?

File: packages/app/src/app/overmind/onInitialize.ts [hunk 7]
```diff
@@ -1,136 +0,0 @@
...
-
-export const onInitializeOvermind = async (
```
Comment (CompuIves) [inline packages/app/src/app/overmind/onInitialize.ts:3]: Oh wow, we don't use this anymore?

File: packages/app/src/app/pages/Sandbox/Editor/Workspace/screens/Deployment/GithubPages/index.tsx [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+  if (
+    template.staticDeployment === false ||
+    template.githubPagesDeploy === false
+  ) {
+    return null;
+  }
+
+  const deploy = () => {
+    track('Deploy Clicked', { provider: 'github' });
```
Comment (CompuIves) [inline packages/app/src/app/pages/Sandbox/Editor/Workspace/screens/Deployment/GithubPages/index.tsx:49]: Nice that you made sure of tracking it! Will be nice to see once we deploy.

**Review by CompuIves — APPROVED** [review]
Very great! I left a couple questions on the code, everything looks good! Will test the flow now...

**Comment by CompuIves** [comment]
Omg okay this flow, this UX, was really great @SaraVieira and @andrasbacsai !! I tried to deploy the new sandbox, and it seems everything went well, but I get a 404 still: https://compuives.github.io/csb-new/. But boy, that flow was a great experience! Seamless!!

**Comment by andrasbacsai** [comment]
Github Pages is slow most of the time. I can access your GH Pages now.

Thanks! @SaraVieira did a pretty good job on the frontend side and helped me debug the weird parts.😄

File: packages/app/src/app/overmind/effects/deployment/githubPages.ts (hunk 3 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/effects/deployment/githubPages.ts:32]: The good old copy and paste ahah

File: packages/app/src/app/overmind/effects/deployment/githubPages.ts (hunk 4 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/effects/deployment/githubPages.ts:98]: We do send the JWT so you can only access your own sandboxes and build those, also sent this to andras for answering

File: packages/app/src/app/overmind/namespaces/deployment/actions.ts (hunk 6 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/deployment/actions.ts:249]: Yes! We started the process

File: packages/app/src/app/overmind/onInitialize.ts (hunk 7 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/onInitialize.ts:3]: Its been moved here `packages/app/src/app/overmind/actions.ts` in the latest release of overmind

File: packages/app/src/app/pages/Sandbox/Editor/Workspace/screens/Deployment/GithubPages/index.tsx (hunk 8 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/Sandbox/Editor/Workspace/screens/Deployment/GithubPages/index.tsx:49]: Same!

**Comment by SaraVieira** [comment]
Yeahhh :(

That's why I have the warning that it can take up to 15m :(

File: packages/app/src/app/overmind/effects/deployment/githubPages.ts (hunk 4 above)
Comment (andrasbacsai) [reply packages/app/src/app/overmind/effects/deployment/githubPages.ts:98]: Good point. I'll check this. Details will be in private.

Comment (andrasbacsai) [reply packages/app/src/app/overmind/effects/deployment/githubPages.ts:98]: What @SaraVieira said, it's true. From that point of view, everybody could build their own sandboxes, nothing more. Our API does authorization.

Comment (CompuIves) [reply packages/app/src/app/overmind/effects/deployment/githubPages.ts:98]: I see, so if I have a build script that runs `ls` and sends the output to one of my servers, I won't be able to see files of other sandboxes that are building at the same time?

Comment (SaraVieira) [reply packages/app/src/app/overmind/effects/deployment/githubPages.ts:98]: Nop, it starts its own service for each sandbox

**Review by christianalfoni — APPROVED** [review]
Awesome work @SaraVieira 😉

**Pull request merged**